### PR TITLE
Mark aliases during scheduling. 

### PR DESCRIPTION
**Pull request by wujingyue**
This is one of the many steps to solve #1401.

Currently, we run alias analysis and mark aliases before segmentation. This can't handle the output-to-output alias patterns we observed in the nanogpt2 benchmark, because some schedulers fork outputs and therefore break the aliasing chain observed before segmentation (and thus before scheduling). https://github.com/NVIDIA/Fuser/issues/1401#issuecomment-1840137527 shows an example for curious minds. 

This PR changes the algorithm so it runs alias analysis both before segmentation and during scheduling. The former is used to change layouts to enable aliases, and the latter is used to detect and mark aliases in each segment. This allows segmentation to split out meta-op-only regions so the rest is easier to optimize. 

This PR doesn't solve the aforementioned output-to-output alias patterns. I'll need to add logic to segment out meta-op-only output-to-output regions in later PRs.

### CODE REVIEW

File: csrc/device_lower/pass/expr_sort.cpp [hunk 1]
```diff
@@ -1749,17 +1749,18 @@ std::vector<Expr*> ExprSegmentationSorter::getExprs() const {
...
+  // this check is needed at this moment.
   if (fusion->isNoOp()) {
     return {};
   }
-  NVF_ERROR(fusion != nullptr);
+
   ExprSegmentationSorter sorter(fusion);
   sorter.sort();
-  auto sorted_exprs = sorter.getExprs();
-  NVF_ERROR(
```
Comment (wujingyue) [inline csrc/device_lower/pass/expr_sort.cpp:1759]: `ExprSegmentationSorter` skips generating code for output aliases, so the expression list can be empty.

**Comment by wujingyue** [comment]
!build

File: csrc/optimization/alias_analysis.cpp [hunk 2]
```diff
@@ -132,6 +132,8 @@ void AliasFinder::handle(const ViewOp* view) {
 
   std::unordered_map<IterDomain*, IterDomain*> in_rfactor_to_out_root =
       PairwiseRootDomainMap(in, out).mapBroadcast(true).mapProducerToConsumer();
+  std::unordered_map<IterDomain*, IterDomain*> out_root_to_in_rfactor =
```
Comment (jjsjann123) [inline csrc/optimization/alias_analysis.cpp:135]: errr. This is coming from #1489 right?

Something doesn't look right on github's code diff.

**Review by jjsjann123 — COMMENTED** [review]
errr. code diff looks wrong on github GUI. I'll try to do a local diff and review that way.

**Comment by jjsjann123** [comment]
Ha I think it's just the conflicts that's making it extra confusing to github. @wujingyue would you mind resolving that first and I'll review afterwards :bow:

**Comment by wujingyue** [comment]
!build

File: csrc/fusion.cpp [hunk 3]
```diff
@@ -21,6 +21,7 @@
 #include <iter_visitor.h>
 #include <kernel.h>
 #include <ops/arith.h>
+#include <optimization/alias_analysis.h>
```
Comment (jjsjann123) [inline csrc/fusion.cpp:24]: is this one an accidental addition?

File: csrc/optimization/alias_analysis.h [hunk 4]
```diff
@@ -22,7 +22,9 @@ struct Layout {
 
   // Returns whether this layout is compliant with `required`. This is
   // uni-directional. For example, `contiguity=[t,t]` is compliant with
-  // `contiguity=[f,f]` but not vice versa.
+  // `contiguity=[f,f]` but not vice versa. As a special case,
+  // an empty `required.allocation` indicates no requirements, i.e., the method
+  // always returns true.
```
Comment (jjsjann123) [inline csrc/optimization/alias_analysis.h:27]: Thanks for the nice comment here!

File: csrc/scheduler/mark_aliases.cpp [hunk 5]
```diff
@@ -0,0 +1,56 @@
...
+       ir_utils::filterByType<TensorView>(fusion->outputs())) {
+    const Val* in = analysis.getAliasedInput(out);
+    if (in == nullptr) {
+      continue;
+    }
+
+    fusion->aliasOutputToInput(
+        out,
+        // NOLINTNEXTLINE(cppcoreguidelines-pro-type-const-cast)
+        const_cast<Val*>(in),
```
Comment (jjsjann123) [inline csrc/scheduler/mark_aliases.cpp:46]: nitpick, not your problem, but I should probably have updated the signature in `aliasOutputToInput` to take `const Val*` instead.

**Review by jjsjann123 — APPROVED** [review]
LGTM

File: csrc/fusion.cpp (hunk 3 above)
Comment (wujingyue) [reply csrc/fusion.cpp]: Done.

**Review by naoyam — APPROVED** [review]
LGTM as well

**Comment by wujingyue** [comment]
!build

**Pull request merged**